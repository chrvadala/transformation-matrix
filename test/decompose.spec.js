/* global describe, it, expect */
import {
  decompose,
  compose,
  identity,
  translate,
  rotate,
  scale
} from '../src'

describe('decompose', () => {
  it('should not return transformations on identity decomposition', () => {
    const decomposition = decompose(identity())
    expect(decomposition).toMatchSnapshot()
  })

  it('should decompose positive translation', () => {
    const composition = compose(
      identity(),
      translate(20, 30)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
  })

  it('should decompose negative translation', () => {
    const composition = compose(
      identity(),
      translate(-20, -30)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
  })

  it('should decompose positive rotation', () => {
    const deg = 40
    const composition = compose(
      identity(),
      rotate(Math.PI * deg / 180)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
  })

  it('should decompose negative rotation', () => {
    const deg = -40
    const composition = compose(
      identity(),
      rotate(Math.PI * deg / 180)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
  })

  it('should decompose positive rotation (3th quadrant)', () => {
    const deg = 260
    const composition = compose(
      identity(),
      rotate(Math.PI * deg / 180)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
    expect(decomposition.rotate).toBeLessThan(0)
  })

  it('should decompose negative rotation (4th quadrant)', () => {
    const deg = 280
    const composition = compose(
      identity(),
      rotate(Math.PI * deg / 180)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
    expect(decomposition.rotate).toBeLessThan(0)
  })

  it('should decompose proportional scaling', () => {
    const composition = compose(
      identity(),
      scale(1.5)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
  })

  it('should decompose distortion scaling', () => {
    const composition = compose(
      identity(),
      scale(1.5, 1)
    )
    const decomposition = decompose(composition)

    expect(decomposition).toMatchSnapshot()
  })
})
