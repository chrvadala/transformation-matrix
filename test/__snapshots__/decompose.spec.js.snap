// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decompose should decompose distortion scaling 1`] = `
Object {
  "rotate": 0,
  "scale": Object {
    "sx": 1.5,
    "sy": 1,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`decompose should decompose negative rotation (4th quadrant) 1`] = `
Object {
  "rotate": -1.396263401595464,
  "scale": Object {
    "sx": 1,
    "sy": 1,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`decompose should decompose negative rotation 1`] = `
Object {
  "rotate": -0.6981317007977318,
  "scale": Object {
    "sx": 0.9999999999999999,
    "sy": 0.9999999999999999,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`decompose should decompose negative translation 1`] = `
Object {
  "rotate": 0,
  "scale": Object {
    "sx": 1,
    "sy": 1,
  },
  "translate": Object {
    "x": -20,
    "y": -30,
  },
}
`;

exports[`decompose should decompose positive rotation (3th quadrant) 1`] = `
Object {
  "rotate": -1.7453292519943295,
  "scale": Object {
    "sx": 0.9999999999999999,
    "sy": 0.9999999999999999,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`decompose should decompose positive rotation 1`] = `
Object {
  "rotate": 0.6981317007977318,
  "scale": Object {
    "sx": 0.9999999999999999,
    "sy": 0.9999999999999999,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`decompose should decompose positive translation 1`] = `
Object {
  "rotate": 0,
  "scale": Object {
    "sx": 1,
    "sy": 1,
  },
  "translate": Object {
    "x": 20,
    "y": 30,
  },
}
`;

exports[`decompose should decompose proportional scaling 1`] = `
Object {
  "rotate": 0,
  "scale": Object {
    "sx": 1.5,
    "sy": 1.5,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;

exports[`decompose should not return transformations on identity decomposition 1`] = `
Object {
  "rotate": 0,
  "scale": Object {
    "sx": 1,
    "sy": 1,
  },
  "translate": Object {
    "x": 0,
    "y": 0,
  },
}
`;
